src/btree/bt_sync.c

```diff
@@ -240,10 +240,15 @@ __sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
 		 * Set the checkpointing flag to block such actions and wait for
 		 * any problematic eviction or page splits to complete.
 		 */
+		WT_ASSERT(session, btree->checkpointing == WT_CKPT_OFF);
+
 		btree->checkpointing = WT_CKPT_PREPARE;
 		(void)__wt_gen_next_drain(session, WT_GEN_EVICT);
 		btree->checkpointing = WT_CKPT_RUNNING;
 
+		/* Note in the session that we are syncing this tree. */
+		session->sync_dhandle = session->dhandle;
+
```

Comment:
Why isn't `WT_SESSION.sync_dhandle` a flag or boolean?

I'm not positive of this, but isn't it the same as holding `WT_SESSION_LOCKED_CHECKPOINT`? (I'm not excited about actually using `WT_SESSION_LOCKED_CHECKPOINT`, but hopefully that helps explain the point.)

Alternatively, it might be more intuitive to have `WT_SESSION.sync_btree`, that makes it more obvious we're referencing the same tree as `WT_BTREE.checkpointing != WT_CKPT_OFF`.

Reply:
Edit: never mind, of course it can't be a boolean.

I still think `WT_SESSION.sync_btree` makes a little more sense than `WT_SESSION.sync_dhandle`.

Reply:
FTR, there is subtlety: if a checkpoint is currently working in ordinary tree T, and as part of that it updates T's metadata, the ordinary rules should be followed for the metadata write.  In particular, there can be a concurrent flush of the metadata (e.g., as part of a `WT_SESSION::create`), and the checkpoint should not be able to evict or split a metadata page in that case.